AttachmentElement cannot be tested with MiniBrowser

kkinnunen-apple opened the pull request:
#### 238cf2fd98964b7cd12b04f476c6739c1f975d03
<pre>
AttachmentElement cannot be tested with MiniBrowser
<a href="https://bugs.webkit.org/show_bug.cgi?id=262686">https://bugs.webkit.org/show_bug.cgi?id=262686</a>
rdar://116509843

Reviewed by Aditya Keerthi.

Add a menu entry to enable AttachmentElement.
The setting takes place for new windows only.

* Tools/MiniBrowser/mac/AppDelegate.m:
(-[BrowserAppDelegate defaultConfiguration]):
* Tools/MiniBrowser/mac/SettingsController.h:
* Tools/MiniBrowser/mac/SettingsController.m:
(-[SettingsController _populateMenu:]):
(-[SettingsController validateMenuItem:]):
(-[SettingsController attachmentElementEnabled]):
(-[SettingsController preferenceValueForAttachmentElementEnabledTag:]):
(-[SettingsController attachmentElementEnabled:]):
(-[SettingsController changeAttachmentElementEnabled:]):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController didChangeSettings]):

Canonical link: <a href="https://commits.webkit.org/269071@main">https://commits.webkit.org/269071@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f4f42baca409d93bd94d1363278433cec572335

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21027 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21408 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22093 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22908 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19562 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21267 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24661 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21593 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20827 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21250 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20987 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18236 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23762 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18147 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19064 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25344 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19227 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19265 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23273 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19811 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16835 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19081 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5139 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23383 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19655 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/MiniBrowser/mac/SettingsController.h:
```diff
@@ -27,6 +27,12 @@
+typedef NS_ENUM(NSInteger, AttachmentElementEnabledState) {
+    AttachmentElementEnabledStateDisabled = 0,
+    AttachmentElementEnabledStateEnabled,
+    AttachmentElementEnabledStateWideLayoutEnabled,
```
@squelart Does anyone use the non-wide layout these days? Wondering if we can simplify this preference, and remove that code.

kkinnunen-apple replied:
I think that decision is such a big decision that it shouldn't affect this review?

kkinnunen-apple replied:
If it's removed, it will be then further removed from MiniBrowser too

squelart replied:
Yes the still-default narrow/legacy layout is still used in Mail when composing -- though eventually that will probably change.
And I believe it's used in Messages as well, but not sure.
So it's good to have the option available in minibrowser.

pxlcoder replied:
Messages doesn't use `<attachment>`, but good to know about Mail!

pxlcoder commented on Tools/MiniBrowser/mac/AppDelegate.m:
```diff
@@ -263,6 +263,9 @@ - (WKWebViewConfiguration *)defaultConfiguration
 
     configuration.suppressesIncrementalRendering = _settingsController.incrementalRenderingSuppressed;
     configuration.websiteDataStore._resourceLoadStatisticsEnabled = _settingsController.resourceLoadStatisticsEnabled;
+    configuration._attachmentElementEnabled = _settingsController.attachmentElementEnabled != AttachmentElementEnabledStateDisabled ? YES : NO;
+    configuration._attachmentWideLayoutEnabled = _settingsController.attachmentElementEnabled == AttachmentElementEnabledStateWideLayoutEnabled ? YES : NO;
```
I think `? YES : NO` is redundant.

pxlcoder commented on Tools/MiniBrowser/mac/SettingsController.m:
```diff
@@ -796,10 +809,43 @@ - (void)changeCustomUserAgent:(id)sender
+- (BOOL)attachmentElementEnabled:(NSMenuItem *)menuItem
+{
+    NSInteger value = [[NSUserDefaults standardUserDefaults] integerForKey:AttachmentElementEnabledPreferenceKey];
+    return [self preferenceValueForAttachmentElementEnabledTag:[menuItem tag]] == value ? YES : NO;
```
I think `? YES : NO` is redundant.

pxlcoder commented on Tools/MiniBrowser/mac/SettingsController.h:
```diff
@@ -55,6 +61,7 @@ extern NSString * const kUserAgentChangedNotificationName;
 @property (nonatomic, readonly) BOOL useSystemAppearance;
 @property (nonatomic, readonly) BOOL dataDetectorsEnabled;
 @property (nonatomic, readonly) BOOL useMockCaptureDevices;
+@property (nonatomic, readonly) AttachmentElementEnabledState attachmentElementEnabled;
```
Nit: Can we move this after the `BOOL` properties?

kkinnunen-apple committed:
AttachmentElement cannot be tested with MiniBrowser
https://bugs.webkit.org/show_bug.cgi?id=262686
rdar://116509843

Reviewed by Aditya Keerthi.

Add a menu entry to enable AttachmentElement.
The setting takes place for new windows only.

* Tools/MiniBrowser/mac/AppDelegate.m:
(-[BrowserAppDelegate defaultConfiguration]):
* Tools/MiniBrowser/mac/SettingsController.h:
* Tools/MiniBrowser/mac/SettingsController.m:
(-[SettingsController _populateMenu:]):
(-[SettingsController validateMenuItem:]):
(-[SettingsController attachmentElementEnabled]):
(-[SettingsController preferenceValueForAttachmentElementEnabledTag:]):
(-[SettingsController attachmentElementEnabled:]):
(-[SettingsController changeAttachmentElementEnabled:]):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController didChangeSettings]):

Canonical link: https://commits.webkit.org/269071@main
